### ISSUE TITLE
Span metrics reported with wrong operation tag

### ISSUE THREAD

**Issue by pnerg** [kamon-io/kamon-akka-http#68]
When using a custom `kamon.akka.http.AkkaHttp.OperationNameGenerator` to render own operation names these get overwritten when the span is reported as a metric.

The span itself is created with the correct name and all sub-spans one runs in the app get the correct parent span. But the request itself will have it's name overwritten when it is completed, then reverting it back to the original URL/path name.

Very simple to re-create
```
package com.area51

import akka.actor.ActorSystem
import akka.http.scaladsl.Http
import akka.http.scaladsl.model.{ContentTypes, HttpEntity, HttpRequest}
import akka.http.scaladsl.server.Directives.{complete, get, pathPrefix, _}
import akka.http.scaladsl.server.Route
import akka.stream.ActorMaterializer
import com.typesafe.config.{Config, ConfigFactory}
import kamon.akka.http.AkkaHttp.OperationNameGenerator
import kamon.metric.PeriodSnapshot
import kamon.{Kamon, MetricReporter}

class SimpleNameGenerator extends OperationNameGenerator {
  override def clientOperationName(request: HttpRequest): String = "MY-CLIENT"
  override def serverOperationName(request: HttpRequest): String = "MY-SERVER"
}

class SimpleMetricReporter extends MetricReporter {
  override def start(): Unit = {}
  override def stop(): Unit = {}
  override def reconfigure(config: Config): Unit = {}
  override def reportPeriodSnapshot(snapshot: PeriodSnapshot): Unit = {
    //snapshot.metrics.histograms
    snapshot.metrics.histograms.filter(_.name.startsWith("span.processing")).foreach(println)
  }
}

object SimpleKamon extends App {

  private val cfg = ConfigFactory.parseString(
    """kamon {
      |  environment.service = "simple-kamon"
      |
      |  akka-http {
      |    name-generator = "com.area51.SimpleNameGenerator"
      |    not-found-operation-name = "unhandled"
      |    add-http-status-code-as-metric-tag = true
      |  }
      |  
      |  span-metric.tags.upstream-service = no
      |  modules {
      |    kamon-system-metrics.auto-start = false
      |    kamon-akka-http.requires-aspectj = true
      |  }
      | span-metrics.scope-spans-to-parent=false
      |  system-metrics {
      |    host.enabled = false
      |    jmx.enabled = false
      |  }
      |  metric.tick-interval = 10 seconds
      |}""".stripMargin
  )
    .withFallback(ConfigFactory.defaultReference())
    .resolve()
  
  private implicit val system = ActorSystem("simple-kamon")
  private implicit val materializer = ActorMaterializer()
  private implicit val executionContext = system.dispatcher
  
  private val route: Route =
    pathPrefix("api") {
      pathPrefix("execute") {
        get {
          println(Kamon.currentSpan().operationName())
          Kamon.buildSpan("read.user.data").withMetricTag("component", "database").start().finish()
          Kamon.buildSpan("update.cache").withMetricTag("component", "in-memory").start().finish()
          complete(HttpEntity(ContentTypes.`text/html(UTF-8)`, s"<h1>HELLO WORLD!</h1>"))
        }
      }
    }
  
  Kamon.reconfigure(cfg)
  Kamon.addReporter(new SimpleMetricReporter())

  Http().bindAndHandle(route, "0.0.0.0", 6969).foreach(_ => println("Started!"))
}
``` 

build.sb as reference
```
val akkaVersion = "2.5.25"
lazy val kamon = project.in(file("kamon"))
  .settings(
    organization := "com.area51",
    scalaVersion := "2.12.10",
    mainClass in (Compile, run) := Some("com.area51.SimpleKamon"),
    libraryDependencies ++= Seq(
      "com.typesafe.akka" %% "akka-actor"   % akkaVersion,
      "com.typesafe.akka" %% "akka-slf4j"   % akkaVersion,
      "com.typesafe.akka" %% "akka-stream"  % akkaVersion,
      "com.typesafe.akka" %% "akka-http-core"  % "10.1.9",
      "com.typesafe.akka" %% "akka-http"       % "10.1.9",
      "io.kamon" %% "kamon-core"           % "1.1.6",
      "io.kamon" %% "kamon-akka-http-2.5"  % "1.1.2",
      "io.kamon" %% "kamon-akka-2.5"       % "1.1.4"
    )
  )
```

Sending a single request to the service one can see these printouts
```
MY-SERVER
MetricDistribution(span.processing-time,Map(operation -> /api/execute, error -> false, span.kind -> server, http.status_code -> 200),MeasurementUnit(Dimension(time),Magnitude(nanoseconds,1.0E-9)),DynamicRange(1,3600000000000,2),kamon.metric.SnapshotCreation$ZigZagCountsDistribution@9e07ae2)
MetricDistribution(span.processing-time,Map(operation -> read.user.data, error -> false, component -> database, parentOperation -> MY-SERVER),MeasurementUnit(Dimension(time),Magnitude(nanoseconds,1.0E-9)),DynamicRange(1,3600000000000,2),kamon.metric.SnapshotCreation$ZigZagCountsDistribution@48b2e947)
MetricDistribution(span.processing-time,Map(operation -> update.cache, error -> false, component -> in-memory, parentOperation -> MY-SERVER),MeasurementUnit(Dimension(time),Magnitude(nanoseconds,1.0E-9)),DynamicRange(1,3600000000000,2),kamon.metric.SnapshotCreation$ZigZagCountsDistribution@5ca82cba)
```
The name of the span is correct but the metric has the wrong _operation_
It all boils down to this line in kamon-akka-http
[ServerRequestInstrumentation:68](https://github.com/kamon-io/kamon-akka-http/blob/maintenance-1.x/kamon-akka-http/src/main/scala/kamon/akka/http/instrumentation/ServerRequestInstrumentation.scala#L68)
It just overwrites the operation name with the original URL.
This causes huge issues as our naming rule is based on knowing our URL patterns, now I get massive cardinality on operation names due to our URLs containing unique identifiers

**Comment by ivantopo** [comment]
Hey @pnerg, probably what we will need to do is something like [this](https://github.com/kamon-io/kamon-akka-http/blob/master/kamon-akka-http/src/main/scala-2.13/kamon/instrumentation/akka/http/AkkaHttpServerInstrumentation.scala#L175-L197) from the latest version: check whether the operation name has been updated by the user and only set the name if it wasn't.

Let me give it a try.

**Comment by ivantopo** [comment]
@pnerg I just published `1.1.2-8c3fc63eb3e7a69a087f773f6157f8c0c3c89dbb` on our [snapshots repo](https://bintray.com/kamon-io/snapshots/kamon-akka-http), could you please give it a try?

**Comment by pnerg** [comment]
Nope, sorry doesn't seem to make any difference.
E.g. using a URL/path like `/api/customer/4321/do-it` will yield metrics like this.
```
MetricDistribution(span.processing-time,Map(operation -> /api/customer/{}/{}, error -> false, span.kind -> server, http.status_code -> 200),MeasurementUnit(Dimension(time),Magnitude(nanoseconds,1.0E-9)),DynamicRange(1,3600000000000,2),kamon.metric.SnapshotCreation$ZigZagCountsDistribution@3e58edc3)
MetricDistribution(span.processing-time,Map(operation -> read.user.data, error -> false, component -> database, parentOperation -> MY-SERVER),MeasurementUnit(Dimension(time),Magnitude(nanoseconds,1.0E-9)),DynamicRange(1,3600000000000,2),kamon.metric.SnapshotCreation$ZigZagCountsDistribution@1ec2b782)
MetricDistribution(span.processing-time,Map(operation -> update.cache, error -> false, component -> in-memory, parentOperation -> MY-SERVER),MeasurementUnit(Dimension(time),Magnitude(nanoseconds,1.0E-9)),DynamicRange(1,3600000000000,2),kamon.metric.SnapshotCreation$ZigZagCountsDistribution@4f9ac445)
```
I expected the operation to be _MY-SERVER_  just as the parentOperation on the metrics for the child spans.

I'm using 
`"io.kamon" %% "kamon-akka-http-2.5"  % "1.1.2-8c3fc63eb3e7a69a087f773f6157f8c0c3c89dbb"`

**Comment by ivantopo** [comment]
@pnerg I just made some extra changes and published `1.1.2-263555c2fea3f51a3305b328a3c489bc9717ca4d` on our Snapshots repo, this one is working nice with the repro you posted above. Please, give it a try and let me know how it goes.

**Comment by pnerg** [comment]
Yup, seems to do the trick...:D
Did a more realistic test with a couple of "real" components to see the outcome, the operation names are now as I'd expect in the Prometheus exporter.
Thx @ivantopo

**Comment by pnerg** [comment]
@ivantopo , could you please merge & release the fixes you made.
I'm currently maintaining an own binary version of this just to get around the problem. Would rather use the official release

**Comment by ivantopo** [comment]
Ok! I'll try to get around it today. Earlier I was working on the equivalent of this issue for 2.0 (you have a PR open in common) so I'm on the topic. Ping you later about this!

**Comment by ivantopo** [comment]
@pnerg released as [1.1.3](https://github.com/kamon-io/kamon-akka-http/releases/tag/v1.1.3)
